cmd/gpu_fakedev/gpu_fakedev.go

```diff
@@ -302,6 +306,114 @@ func getOptions(name string) genOptions {
 	return opts
 }
 
+func makeXelinkSideCar(topology string, gpus int, tiles int, connections string) {
+	if topology != fullyConnected {
+		log.Printf("XELINK: generate xelink sidecar label file, using (GPUs: %d, Tiles: %d, Connections: %s)", gpus, tiles, connections)
+	} else {
+		log.Printf("XELINK: generate xelink sidecar label file, using (GPUs: %d, Tiles: %d, Topology: %s)", gpus, tiles, topology)
+	}
+
+	smap := ""
+
+	if topology == fullyConnected {
+		// create 2d slice with zeros
+		var cmap = make([][]int, 0)
+
+		var total = gpus * tiles
+
+		for y := 0; y < total; y++ {
+			var xmap = make([]int, 0)
+
+			for x := 0; x < total; x++ {
+				if x != y {
+					xmap = append(xmap, 1)
+				} else {
+					xmap = append(xmap, 0)
+				}
+			}
+
+			cmap = append(cmap, [][]int{xmap}...)
+		}
+
+		// filter double connections
+		for y := 0; y < total; y++ {
+			for x := 0; x < total; x++ {
+				if cmap[y][x] == cmap[x][y] {
+					cmap[x][y] = 0
+				}
+			}
+		}
+
+		// make connection string
+		smap = buildConnectionList(gpus, tiles, cmap)
+	} else {
+		smap = connections
+	}
+
+	saveSideCarFile(smap)
+}
+
+func buildConnectionList(gpus int, tiles int, cmap [][]int) string {
+	smap := ""
+
+	for gpuy := 0; gpuy < gpus; gpuy++ {
+		for tiley := 0; tiley < tiles; tiley++ {
+			y := gpuy*tiles + tiley
+
+			for gpux := 0; gpux < gpus; gpux++ {
+				for tilex := 0; tilex < tiles; tilex++ {
+					x := gpux*tiles + tilex
+					if cmap[y][x] == 1 {
+						smap = fmt.Sprintf("%s_%d.%d-%d.%d", smap, gpuy, tiley, gpux, tilex)
+					}
+				}
+			}
+		}
+	}
+
+	if smap != "" {
+		smap = smap[1:]
+	}
+
+	return smap
+}
+
+func saveSideCarFile(connections string) {
+	f, err := os.Create("xpum-sidecar-labels.txt")
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	if len(connections) > maxNfdLileLen {
+		line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections[:maxNfdLileLen])
+		fmt.Println(line)
+
+		if _, err := f.WriteString(line + "\n"); err != nil {
+			panic(err)
+		}
+
+		index := 2
+
+		for i := maxNfdLileLen; i < len(connections); i += (maxNfdLileLen - 1) {
+			line := fmt.Sprintf("xpumanager.intel.com/xe-links%d=Z%s", index, connections[i:int(math.Min(float64(len(connections)), float64(i+maxNfdLileLen-1)))])
+			fmt.Println(line)
+
+			if _, err := f.WriteString(line + "\n"); err != nil {
+				panic(err)
+			}
+			index++
+		}
+	} else {
+		line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections[:int(math.Min(float64(len(connections)), float64(maxNfdLileLen)))])
```

Comment:
Shouldn't the `connections` be written as is in this case? I don't understand the `int(math.Min(float64` parts.